### PR TITLE
[DeadStoreElimination] Refactor out `pushMemUses`, drop dead check (NFC)

### PR DESCRIPTION
**Pull request by antoniofrighetto**

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 1]
```diff
@@ -1162,21 +1172,11 @@ struct DSEState {
                       << *Def->getMemoryInst()
                       << ") is at the end the function \n");
 
-    auto MaybeLoc = getLocForWrite(Def->getMemoryInst());
-    if (!MaybeLoc) {
-      LLVM_DEBUG(dbgs() << "  ... could not get location for write.\n");
-      return false;
-    }
-
+    MemoryLocation DefLoc = *getLocForWrite(Def->getMemoryInst());
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1175]: I would feel more comfortable if you passed in this MemoryLocation already computed in the caller instead. That makes it clearer that this really cannot fail.

Comment (antoniofrighetto) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1175]: We still want to push the memory access in the worklist though, so we would be still passing the MemoryDef, right? Felt a bit cleaner to pass only one the MemoryDef.

Comment (nikic) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1175]: Yes, you should pass both the MemoryDef and MemoryLocation.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 2]
```diff
@@ -1157,26 +1167,14 @@ struct DSEState {
   }
 
   /// Returns true if \p Def is not read before returning from the function.
-  bool isWriteAtEndOfFunction(MemoryDef *Def) {
+  bool isWriteAtEndOfFunction(MemoryDef *Def, MemoryLocation DefLoc) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1170]:
```suggestion
  bool isWriteAtEndOfFunction(MemoryDef *Def, const MemoryLocation &DefLoc) {
```

Comment (antoniofrighetto) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Sorry, shouldn't commit after 10pm.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
FYI this PR is not NFC. See https://github.com/dtcxzyw/llvm-opt-benchmark/issues/930.
Hopefully this patch doesn't cause miscompilation :)

**Comment by nikic** [comment]
Yes, it's not NFC, but I think the new code is more correct (previously I think we had some duplicates in the worklist, which made it more likely to hit the cutoff point).

**Comment by antoniofrighetto** [comment]
Right, this is not a NFC, since we were previously not taking into account all the visited MemoryAccess in `isWriteAtEndOfFunction`. From a quick glance at the diffs of a few samples, this should be OK.

**Comment by fhahn** [comment]
@antoniofrighetto might be good to add a test case from https://github.com/dtcxzyw/llvm-opt-benchmark/issues/930 if possible

**Comment by fhahn** [comment]
> > @antoniofrighetto might be good to add a test case from [dtcxzyw/llvm-opt-benchmark#930](https://github.com/dtcxzyw/llvm-opt-benchmark/issues/930) if possible
> 
> Added in [10d720b](https://github.com/llvm/llvm-project/commit/10d720b5b4383c9900caaed7edf38ae31e8b9ac4), thanks!

Many thanks!

**Pull request merged**